pkg/webhooks/tensorflow/tfjob_webhook.go

```diff
@@ -0,0 +1,121 @@
+/*
+Copyright 2024 The Kubeflow Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package tensorflow
+
+import (
+	"context"
+	"fmt"
+	"strings"
+
+	apimachineryvalidation "k8s.io/apimachinery/pkg/api/validation"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/util/validation/field"
+	"k8s.io/klog/v2"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/webhook"
+	"sigs.k8s.io/controller-runtime/pkg/webhook/admission"
+
+	trainingoperator "github.com/kubeflow/training-operator/pkg/apis/kubeflow.org/v1"
+)
+
+var (
+	specPath          = field.NewPath("spec")
+	tfReplicaSpecPath = specPath.Child("tfReplicaSpecs")
+)
+
+type Webhook struct{}
+
+func SetupWebhook(mgr ctrl.Manager) error {
+	return ctrl.NewWebhookManagedBy(mgr).
+		For(&trainingoperator.TFJob{}).
+		WithValidator(&Webhook{}).
+		Complete()
+}
+
+// +kubebuilder:webhook:path=/validate-kubeflow-org-v1-tfjob,mutating=false,failurePolicy=fail,sideEffects=None,groups=kubeflow.org,resources=tfjobs,verbs=create;update,versions=v1,name=validator.tfjob.training-operator.kubeflow.org,admissionReviewVersions=v1
+
+var _ webhook.CustomValidator = &Webhook{}
+
+func (w *Webhook) ValidateCreate(ctx context.Context, obj runtime.Object) (admission.Warnings, error) {
+	job := obj.(*trainingoperator.TFJob)
+	log := ctrl.LoggerFrom(ctx).WithName("tfjob-webhook")
+	log.V(5).Info("Validating create", "TFJob", klog.KObj(job))
+	return nil, validateTFJob(job).ToAggregate()
+}
+
+func (w *Webhook) ValidateUpdate(ctx context.Context, _, newObj runtime.Object) (admission.Warnings, error) {
+	job := newObj.(*trainingoperator.TFJob)
+	log := ctrl.LoggerFrom(ctx).WithName("tfjob-webhook")
+	log.V(5).Info("Validating update", "NewTFJob", klog.KObj(job))
+	return nil, validateTFJob(job).ToAggregate()
+}
+
+func (w *Webhook) ValidateDelete(context.Context, runtime.Object) (admission.Warnings, error) {
+	return nil, nil
+}
+
+func validateTFJob(job *trainingoperator.TFJob) field.ErrorList {
+	var allErrs field.ErrorList
+	if errors := apimachineryvalidation.NameIsDNS1035Label(job.Name, false); len(errors) != 0 {
+		allErrs = append(allErrs, field.Invalid(field.NewPath("metadata").Child("name"), job.Name, fmt.Sprintf("should match: %v", strings.Join(errors, ","))))
+	}
```

Comment:
Should we move this validation to the helper function in `/webhooks/utils/utils.go` since this check is common to all Framework ?

Reply:
Yes, we can commonize these functions, but as I mentioned in the description, commonizing in a separate PR would be better. Because the minimum validation change could prove that webhook validations would not bring any bugs/breaks.

> Although we potentially commonize podTemplate validations, I leave such commonize in the follow-ups.

Reply:
Sure, that sounds good @tenzen-y!
We might need to track it in the separate issue.

Reply:
Sure

Reply:
Created: #2054